react/finger-exercises/redux/02 - react-redux/src/redux/book/reducer.js

```diff
@@ -3,21 +3,45 @@ import { actions } from './actions';
 const initialState = {
   books: [],
   bookSelected: [],
-  originalData: []
+  originalData: [],
+  bookSearch: null
 };
 
 function reducer(state = initialState, action) {
   switch (action.type) {
-    case actions.GET_BOOKS: // TODO to implement the logic
-      return { ...state };
-    case actions.ADD_TO_CART: // TODO to implement the logic
-      return { ...state };
-    case actions.ADD_ITEM: // TODO to implement the logic
-      return { ...state };
-    case actions.REMOVE_ITEM: // TODO to implement the logic
-      return { ...state };
-    case actions.SEARCH_ITEM: // TODO to implement the logic
-      return { ...state };
+    case actions.GET_BOOKS:
+      return {
+        ...state,
+        books: [...action.payload]
+      };
+    case actions.ADD_TO_CART:
+      return {
+        ...state,
+        bookSelected: [...state.bookSelected, action.payload]
+      };
+    case actions.ADD_ITEM:
+      return {
+        ...state,
+        bookSelected: [
+          ...state.bookSelected.map(item =>
+            item.id === action.payload ? { ...item, quantity: item.quantity + 1 } : item
+          )
+        ]
+      };
+    case actions.REMOVE_ITEM:
+      return {
+        ...state,
+        bookSelected: [
+          ...state.bookSelected
+            .map(item => (item.id === action.payload ? { ...item, quantity: item.quantity - 1 } : item))
+            .filter(item => item.quantity !== 0)
+        ]
```

Comment:
A good practice is to have all the logic in the action. And in the reducer you just set the new value.